pusher/pusher.go

```diff
@@ -0,0 +1,134 @@
+package pusher
+
+import (
+	"bytes"
+	"crypto/sha1"
+	"fmt"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/kochman/hotshots/config"
+	"github.com/kochman/hotshots/log"
+)
+
+// Pusher is responsible for uploading photos from a camera to a remote location.
+// It only uploads photos that don't exist remotely, in an effort to reduce bandwidth usage.
+type Pusher struct {
+	cfg               config.Config
+	cameraService     cameraService
+	filenameToPhotoID map[string]string
+	photoService      photoService
+}
+
+// New creates a new Pusher.
+func New(cfg *config.Config) (*Pusher, error) {
+	p := &Pusher{
+		cfg:               *cfg,
+		cameraService:     newLocalCamera(),
+		filenameToPhotoID: map[string]string{},
+		photoService: &remoteAPI{
+			url: cfg.ServerURL,
+		},
+	}
+
+	return p, nil
+}
+
+// Run runs the Pusher's upload functionality in a loop forever.
+func (p *Pusher) Run() {
+	ticker := time.NewTicker(time.Second)
```

Comment:
Do you want to lower this ticker to save battery?

Reply:
Yes. I'll add a config field for this and default it to five seconds or so.

I think we could be a lot smarter about it. For example, if no photos have been taken in the last few min, only check every ten seconds. But if there's continuous shooting, then reduce the interval to a second or two.

Reply:
There are also libgphoto2 APIs for getting notifications of new photos on the camera. I don't know if gphoto2go wraps them, but I'll take a look.